addOns/authhelper/src/main/java/org/zaproxy/addon/authhelper/spiderajax/ClientScriptBasedAuthHandler.java

```diff
@@ -0,0 +1,115 @@
+import org.zaproxy.zap.extension.selenium.ExtensionSelenium;
+import org.zaproxy.zap.extension.selenium.SeleniumScriptUtils;
+import org.zaproxy.zap.extension.spiderAjax.AuthenticationHandler;
+import org.zaproxy.zap.model.Context;
+import org.zaproxy.zap.users.User;
+import org.zaproxy.zest.impl.ZestBasicRunner;
+
+public class ClientScriptBasedAuthHandler implements AuthenticationHandler {
+
+    private static final Logger LOGGER = LogManager.getLogger(ClientScriptBasedAuthHandler.class);
+
+    private BrowserHook browserHook;
+    private static ZestAuthRunner zestRunner;
+
+    @Override
+    public boolean enableAuthentication(User user) {
+        Context context = user.getContext();
+        if (context.getAuthenticationMethod()
+                instanceof
+                ClientScriptBasedAuthenticationMethodType.ClientScriptBasedAuthenticationMethod) {
+
+            if (browserHook != null) {
+                throw new IllegalStateException("BrowserHook already enabled");
+            }
+            browserHook = new AuthenticationBrowserHook(context, user);
+
+            AuthUtils.getExtension(ExtensionSelenium.class).registerBrowserHook(browserHook);
+
+            return true;
+        }
+        return false;
+    }
+
+    @Override
+    public boolean disableAuthentication(User user) {
+        if (browserHook != null) {
+            AuthUtils.getExtension(ExtensionSelenium.class).deregisterBrowserHook(browserHook);
+            browserHook = null;
+            return true;
+        }
+        return false;
+    }
+
+    static class AuthenticationBrowserHook implements BrowserHook {
+
+        private ClientScriptBasedAuthenticationMethod csaMethod;
+        private Context context;
+
+        AuthenticationBrowserHook(Context context, User user) {
+            this.context = context;
+            AuthenticationMethod method = context.getAuthenticationMethod();
+            if (!(method instanceof ClientScriptBasedAuthenticationMethod)) {
+                throw new IllegalStateException("Unsupported method " + method.getType().getName());
+            }
+            csaMethod = (ClientScriptBasedAuthenticationMethod) method;
+        }
+
+        private static ZestBasicRunner getZestRunner(WebDriver webDriver) {
+            if (zestRunner == null) {
+                zestRunner = new ZestAuthRunner();
+                // Always proxy via ZAP
+                ServerInfo mainProxyInfo =
+                        AuthUtils.getExtension(ExtensionNetwork.class).getMainProxyServerInfo();
+                zestRunner.setProxy(mainProxyInfo.getAddress(), mainProxyInfo.getPort());
+            }
+            zestRunner.setWebDriver(webDriver);
+            return zestRunner;
+        }
+
+        @Override
+        public void browserLaunched(SeleniumScriptUtils ssUtils) {
+            ZestBasicRunner runner = getZestRunner(ssUtils.getWebDriver());
+            runner.addWebDriver("ClientScriptBasedAuthHandler", ssUtils.getWebDriver());
```

Comment:
Is this needed? Are we using this custom window anywhere?

Reply:
Try now, I don't think we are.